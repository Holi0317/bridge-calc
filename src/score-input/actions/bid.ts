import { IPlayerMap } from "../../types";
import { ActionTypes } from "../../action-types";

export interface IBidAction {
  type: ActionTypes.BID;
  payload?: IPlayerMap<number>;
}

/**
 * End bidding process in bridge game
 * @param payload - A map that maps player ID to their bid choice.
 * If undefined, bid property in currentGame state will be used
 */
export function bidAction(payload?: IPlayerMap<number>): IBidAction {
  return { type: ActionTypes.BID, payload };
}
