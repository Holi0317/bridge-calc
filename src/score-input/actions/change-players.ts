import { IPlayerMap } from "../../types";
import { ActionTypes } from "../../action-types";

export interface IChangePlayersAction {
  type: ActionTypes.CHANGE_PLAYERS;
  newNames: IPlayerMap<string>;
  maker: string;
  rounds: number;
  time: number;
}

/**
 * Change players of current game.
 * Change can include: Change name, add player or remove player.
 * After this action is dispatched, the stage will revert to waitingBid if original stage is waitingWin. (To avoid invalid bid after changing player)
 *
 * This action can also be used to change maker. Just pass in original name map, original rounds
 * and specify new maker in `maker` property.
 *
 * This action can also be used to change number of rounds. Similar to change maker, only update rounds property.
 *
 * @param newNames - New PlayerMap for the names.
 * Order of this map represent the new player order.
 * If new player is added, generate a new unique ID for him/her.
 *
 * @param maker - As order may have changed, new maker must be selected.
 * This property represent the ID of new maker.
 * The player order will be sorted by this property and newNames property.
 * This can also be used to change maker.
 *
 * @param rounds - Number of rounds after dispatching this action.
 * If the given rounds is less than current round, the game will end immediately.
 * The above case is considered as a bug. Prevention of this should be done on the caller side.
 */
export function changePlayersAction(
  newNames: IPlayerMap<string>,
  maker: string,
  rounds: number
): IChangePlayersAction {
  return {
    type: ActionTypes.CHANGE_PLAYERS,
    newNames,
    maker,
    rounds,
    time: new Date().getTime()
  };
}
