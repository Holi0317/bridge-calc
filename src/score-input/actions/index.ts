export * from "./bid";
export * from "./change-players";
export * from "./set-bid";
export * from "./set-win";
export * from "./skip";
export * from "./start";
export * from "./undo";
export * from "./win";
export * from "./replace-current-game";
