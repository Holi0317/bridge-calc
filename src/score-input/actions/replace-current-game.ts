import { createAction } from "typesafe-actions";
import { GameState } from "../reducer";
import { ActionTypes } from "../../action-types";

export const replaceCurrentGameAction = createAction(
  ActionTypes.REPLACE_CURRENT_GAME,
  (payload: GameState) => payload
)();
