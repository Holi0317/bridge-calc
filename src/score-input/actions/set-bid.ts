import { IPlayerMap } from "../../types";
import { ActionTypes } from "../../action-types";

export interface ISetBidAction {
  type: ActionTypes.SET_BID;
  payload: IPlayerMap<number>;
}

/**
 * Set the bid property in current game cache.
 * Not to be confused with BID action.
 *
 * @param payload - A map that maps player ID to their bid choice
 */
export function setBidAction(payload: IPlayerMap<number>): ISetBidAction {
  return { type: ActionTypes.SET_BID, payload };
}
