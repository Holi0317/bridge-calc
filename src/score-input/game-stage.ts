/**
 * Enum of game stages
 */
export enum GameStage {
  /**
   * Waiting for bid result in this round.
   */
  waitingBid = "waitingBid",
  /**
   * Waiting for win result in this round.
   */
  waitingWin = "waitingWin",
  /**
   * Whole game has ended.
   */
  ended = "ended"
}
