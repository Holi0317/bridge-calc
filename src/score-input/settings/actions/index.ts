import { InitSettingsAction } from "./init-settings";
import { SetMakerAction } from "./set-maker";
import { SetNamesAction } from "./set-names";
import { AddNameAction } from "./add-name";
import { ToggleExpandAction } from "./toggle-expand";

export type GameSettingsActions =
  | InitSettingsAction
  | SetMakerAction
  | SetNamesAction
  | AddNameAction
  | ToggleExpandAction;
