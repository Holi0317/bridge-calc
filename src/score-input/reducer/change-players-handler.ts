import mapValues from 'lodash-es/mapValues'
import range from 'lodash-es/range'
import {toEndedState, toWaitingBidState} from './converter'
import {toFront} from '../../utils'
import {IEndedState, IWaitingBidState, IWaitingWinState} from './types'
import {IChangePlayersAction} from '../actions/change-players'

export function changePlayersHandler(rawState: IWaitingBidState | IWaitingWinState, {newNames, rounds, maker, time}: IChangePlayersAction): IWaitingBidState | IEndedState {
  // Short circuit. When action.rounds is less than current round
  if (rounds < rawState.currentRound) {
    // tslint:disable-next-line: no-shadowed-variable
    const state = toEndedState(rawState, time)
    state.names = newNames
    state.rounds = rounds
    state.scores = mapValues(state.scores, (score: number[]) => score.slice(0, rounds))
    return state
  }

  const state = toWaitingBidState(rawState)
  state.rounds = rounds

  // Set currentPlayerOrder
  const newPlayerIDList = Object.keys(newNames)
  state.currentPlayerOrder = toFront(newPlayerIDList, newPlayerIDList.indexOf(maker))

  // Change scores
  const oldScores = state.scores
  const freshScores = range(state.currentRound - 1).fill(0) // Score for new players. I am terrible at naming.
  state.scores = mapValues(newNames, (name: string, ID: string): number[] => (
    (ID in oldScores)
      ? oldScores[ID]
      : freshScores
  ))

  // Change bid
  const oldBid = state.bid
  state.bid = mapValues(newNames, (name: string, ID: string): number => (
    (ID in oldBid)
      ? oldBid[ID]
      : 0
  ))

  // Change names field lastly
  state.names = newNames

  return state
}
