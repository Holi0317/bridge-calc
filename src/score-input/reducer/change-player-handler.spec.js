import {changePlayersHandler} from './change-players-handler'
import {changePlayersAction} from '../actions/change-players'
import {genMap, waitingBidState, waitingWinState} from '../../../test-fixtures/current-game-states'
import * as lolex from 'lolex'
import {GameStage} from '../game-stage'

test('it should no-op when new player map is identical to original one', () => {
  const state = {
    ...waitingBidState
  }
  const newNames = genMap('John', 'Mary', 'Henry', 'Joe')
  const action = changePlayersAction(newNames, 'a', 13)
  const expected = {
    ...waitingBidState
  }
  const actual = changePlayersHandler(state, action)
  expect(actual).toEqual(expected)
})

test('it should revert waiting win state to waiting bid state', () => {
  const state = {
    ...waitingWinState
  }
  const newNames = genMap('John', 'Mary', 'Henry', 'Joe')
  const action = changePlayersAction(newNames, 'a', 13)
  const expected = {
    ...waitingBidState
  }
  const actual = changePlayersHandler(state, action)
  expect(actual).toEqual(expected)
})

test("it should change player's names from given map", () => {
  const expected = {
    ...waitingBidState,
    names: genMap('John', 'DPGJW', 'Henry', 'Joe')
  }
  const state = {
    ...waitingBidState
  }
  const newNames = genMap('John', 'DPGJW', 'Henry', 'Joe')
  const action = changePlayersAction(newNames, 'a', 13)
  const actual = changePlayersHandler(state, action)
  expect(actual).toEqual(expected)
})

test('it should change all maps order from given map (For no addition nor deletion)', () => {
  function genReorderedMap(b, c, d, a) {
    return {b, c, d, a}
  }

  const expected = {
    ...waitingBidState,
    names: genReorderedMap('Mary', 'Henry', 'Joe', 'John'),
    scores: genReorderedMap([], [], [], []),
    currentPlayerOrder: ['b', 'c', 'd', 'a'],
    bid: genReorderedMap(1, 0, 1, 0)
  }
  const state = {
    ...waitingBidState,
    bid: genMap(0, 1, 0, 1)
  }
  const newNames = genReorderedMap('Mary', 'Henry', 'Joe', 'John')
  const action = changePlayersAction(newNames, 'b', 13)
  const actual = changePlayersHandler(state, action)
  expect(actual).toEqual(expected)
})

test('it should re-order players by given maker', () => {
  const expected = {
    ...waitingBidState,
    currentPlayerOrder: ['c', 'd', 'a', 'b']
  }
  const state = {
    ...waitingBidState
  }

  const newNames = genMap('John', 'Mary', 'Henry', 'Joe')
  const action = changePlayersAction(newNames, 'c', 13)
  const actual = changePlayersHandler(state, action)
  expect(actual).toEqual(expected)
})

test('it should change rounds from given payload', () => {
  const expected = {
    ...waitingBidState,
    rounds: 12
  }
  const state = {
    ...waitingBidState
  }
  const newNames = genMap('John', 'Mary', 'Henry', 'Joe')
  const action = changePlayersAction(newNames, 'a', 12)
  const actual = changePlayersHandler(state, action)
  expect(actual).toEqual(expected)
})

test('it should remove all data from removed player', () => {
  function genSmallMap(a, b, c) {
    return {a, b, c}
  }

  const expected = {
    ...waitingBidState,
    names: genSmallMap('John', 'Mary', 'Henry'),
    scores: genSmallMap([], [], []),
    currentPlayerOrder: ['a', 'b', 'c'],
    bid: genSmallMap(0, 1, 0)
  }
  const state = {
    ...waitingBidState,
    bid: genMap(0, 1, 0, 1)
  }
  const newNames = genSmallMap('John', 'Mary', 'Henry')
  const action = changePlayersAction(newNames, 'a', 13)
  const actual = changePlayersHandler(state, action)
  expect(actual).toEqual(expected)
})

test('it should assign 0 mark for ended rounds of new player', () => {
  function genNewMap(a, b, c, d, e) {
    return {b, c, d, a, e}
  }

  const expected = {
    ...waitingBidState,
    currentRound: 2,
    names: genNewMap('John', 'Mary', 'Henry', 'Joe', 'DPGJW'),
    currentPlayerOrder: ['b', 'c', 'd', 'a', 'e'],
    scores: genNewMap([10], [11], [10], [-1], [0]),
    bid: genNewMap(0, 0, 0, 0, 0)
  }
  const state = {
    ...waitingBidState,
    currentRound: 2,
    currentPlayerOrder: ['b', 'c', 'd', 'a'],
    scores: genMap([10], [11], [10], [-1])
  }
  const newNames = genNewMap('John', 'Mary', 'Henry', 'Joe', 'DPGJW')
  const action = changePlayersAction(newNames, 'b', 13)
  const actual = changePlayersHandler(state, action)
  expect(actual).toEqual(expected)
})

test('it should end game when given rounds is less than current round', () => {
  const clock = lolex.install()

  // Short hand for score
  const s = [0, 0]
  const state = {
    ...waitingBidState,
    currentRound: 3,
    currentPlayerOrder: ['c', 'd', 'a', 'b'],
    scores: genMap(s, s, s, s)
  }
  const newNames = genMap('Mary', 'John', 'Henry', 'Joe')
  const action = changePlayersAction(newNames, 'a', 1)
  const actual = changePlayersHandler(state, action)

  expect(actual.stage).toEqual(GameStage.ended)
  expect(actual).toMatchSnapshot()

  clock.uninstall()
})
