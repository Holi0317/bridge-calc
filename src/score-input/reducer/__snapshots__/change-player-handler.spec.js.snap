// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should end game when given rounds is less than current round 1`] = `
Object {
  "endTime": 0,
  "id": "testing",
  "names": Object {
    "a": "Mary",
    "b": "John",
    "c": "Henry",
    "d": "Joe",
  },
  "rounds": 1,
  "scores": Object {
    "a": Array [
      0,
    ],
    "b": Array [
      0,
    ],
    "c": Array [
      0,
    ],
    "d": Array [
      0,
    ],
  },
  "stage": "ended",
  "startTime": 0,
}
`;
